### PR TITLE
evaluate logic ops lazily

### PR DESCRIPTION
**Pull request by BusyJay**
ref pingcap/tidb#2563

@shenli @siddontang @disksing PTAL

### CODE REVIEW

File: src/util/xeval/evaluator.rs [hunk 1]
```diff
@@ -407,6 +381,46 @@ impl Evaluator {
             Ok(left)
         }
     }
+
+    fn eval_logic<F>(&mut self,
+                     ctx: &EvalContext,
+                     expr: &Expr,
+                     break_res: Option<bool>,
+                     logic_eval: F)
```
Comment (siddontang) [inline src/util/xeval/evaluator.rs]: `logic_eval` confuses me because the function name is `eval_logic`.

**Comment by siddontang** [comment]
Miss a test with left is 0 and right is null for `and` and `or`.

**Comment by disksing** [comment]
For me personally, introducing `eval_logic` makes it hard to read. I think they are suitable to be implemented separately. FYI, `if_null` looks much more simple https://github.com/pingcap/tikv/blob/master/src/util/xeval/evaluator.rs#L377

**Comment by BusyJay** [comment]
cc @shenli

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @disksing

**Comment by BusyJay** [comment]
@coocood PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**